fix(discover) Add missing event fields

evanh opened the pull request:
Now if one of these fields is used in a query against the transactions table,
they will get properly NULL'd out instead of being passed through unchanged.

evanh committed:
fix(discover) Add missing event fields

Now if one of these fields is used in a query against the transactions table,
they will get properly NULL'd out instead of being passed through unchanged.

fpacifici commented on snuba/datasets/discover.py:
```diff
@@ -220,6 +221,19 @@ def __init__(self) -> None:
+                ("device_orientation", Nullable(String())),
+                ("device_simulator", Nullable(UInt(8))),
+                ("device_online", Nullable(UInt(8))),
+                ("device_charging", Nullable(UInt(8))),
```
All these fields are promoted contexts, they should never be queries directly (https://github.com/getsentry/snuba/blob/master/snuba/datasets/events.py#L142-L156) they should always be queried as `contexts['device.name']`. 

Do you have an example of query where these columns are queried directly ?

lynnagara replied:
As far as I know nothing in sentry ever referenced a column in the form "contexts[..]"

fpacifici replied:
`If all these values are available through contexts in the errors, events and transactions datasets, we could update sentry to use contexts[] instead of the concrete fields.` 

Can I consider this a plan for "we are going to stop referencing directly the promoted columns" :) ?

`We have been using the concrete columns where possible as they generally perform better.`
This is not 100% accurate:
1) if you query a tag/context and that one is promoted (and it is not one of the exceptions in the next point) then there is no performance gain whatsoever since Snuba translates the query into the concrete column. Though using the tags[]/contests[] column make things easier to understand because you do not have to dela with different data types and you do not have to deal with two ways to do the same thing.
2) there are a few columns in transactions/discover where this is not true, like the trace_id because the context is formatted as string and the column is formatted as UUID. We need some additional query processing capability before we are able to avoid this. But that applies only to a handful of columns in transactions.
3) if you have a tags[]/contexts[] in the query and that is not promoted, then you do not even have a concrete column to query.

So, unless you are looking for a transactions specific columns, go for tags/contexts. And even in the case of transactions, we are going to fix it.

evanh replied:
@lynnagara We are using contexts for a couple transaction fields in sentry. See https://github.com/getsentry/sentry/blob/master/src/sentry/snuba/events.py#L169

@fpacifici So is the ideal case here to switch to using context[...] syntax for all queries against promoted columns? That's easily done I think. How do I tell which columns should be queried using the context/tag?

markstory replied:
@evanh All of the `device.*`, `sdk.*`, `http.*`, `geo.*`, `os.*` fields come from contexts.

markstory replied:
My thinking of 'performs better' was that clickhouse performs better using the concrete columns over unpacking the contexts arrays. Given that snuba can upgrade a `contexts[]` field to the concrete column transparently using contexts would be simpler overall.

fpacifici replied:
@evanh `context[...] syntax for all queries against promoted columns?` . contexts[...] for promoted contexts and tags[...] for promoted tags. They both exists.

both, your call, let me know if you want to migrate to contexts instead of adding these columns or if you need this as a stop gap.

evanh replied:
@fpacifici I am almost done with a PR shifting all the fields Mark mentioned to use contexts[...], but which fields are promoted tags and should use tags[...]?

evanh committed:
more stuff

evanh committed:
Merge branch 'master' into evanh/fix/device_charging_field

evanh committed:
fix(contexts) Remove dots from column to check for promotion

When we were checking to see if a column should be promoted, we weren't removing
the dots from the column name, which meant we might miss some column conversions

evanh committed:
Merge branch 'master' into evanh/fix/device_charging_field

fpacifici commented on snuba/datasets/tags_column_processor.py:
```diff
@@ -148,6 +148,12 @@ def __tag_expr(self, parsed_col: ParsedNestedColumn, table_alias: str = "",) ->
+                tag_name.replace(".", "_"), tag_name
+            )
+            if actual_tag in self.__promoted_columns[col]:
+                return qualified_column(self.__string_col(actual_tag), table_alias)
```
This is not the right way to fix this behavior. 
Promoted columns having the same name of the original tag/context with dots replaced by underscore is a coincidence of the events and transactions tables. This is not supposed to be the default behavior applied during tag processing.

The right place to do this is in the events dataset itself:
https://github.com/getsentry/snuba/blob/master/snuba/datasets/events.py#L372-L377
Where the tags in the form `tag_name.tag_inner_name` are mapped into `tag_name_tag_inner_name`. 
You should add a similar mapping for contexts.

evanh committed:
Merge branch 'master' into evanh/fix/device_charging_field

evanh committed:
address feedback

fpacifici reviewed (commented):
Seems ok, just one last clarification, this will return wrong types till I don't fix the type casting in the tags processor. Are we on the same page on this?
I'd rather have that done first so we are never inconsistent in the response. I will try to do that today, should I not manage to do that I will accept this.

evanh commented:
@fpacifici Yes I understand that this doesn't fix the incorrect typing issue. Doing that fix first makes sense to me.

evanh committed:
Merge branch 'master' into evanh/fix/device_charging_field

evanh committed:
add test for boolean field

fpacifici commented on snuba/datasets/discover.py:
```diff
@@ -47,7 +47,7 @@ def detect_table(
         for idx, condition in enumerate(conditions):
             if is_condition(condition):
-                if tuple(condition) == ("type", "!=", "transaction"):
+                if tuple(condition) == ("type", "=", "error"):
```
I think there are other event types that need to go to the events table (csp, default, etc.) What's the reason for changing this condition ?

markstory replied:
We would either need cases for csp, security and default types or revert this change.

evanh replied:
I changed this for two reasons:

1) The check is broken. We will never match `("type", "!=", "transaction")` because the condition sent from sentry looks like `[[["isNull",["type"]],"=",1],["type","!=","transaction"]]`. 
2) It is useful to have some guarantee which dataset discover queries ultimately run against. Right now this works with transactions but not errors, so having this here actually fixes a number of queries that have `event.type:error` but are not running against the events dataset.

Also, this is a short circuit check for when people specify a specific event. If they specify an event type not in this condition, it will just continue on using the checks below. We don't need new cases here.

fpacifici replied:
" this is a short circuit check for when people specify a specific event. If they specify an event type not in this condition, it will just continue on using the checks below"

I don't know whether the intention of this condition is such. There are a few other conditions between this one and the final return EVENTS. That seems fine to me but I'll let @markstory  confirm. Anyway, since the condition you mention is actually broken today, I guess this does not matter.

markstory replied:
If this condition is not working, how does the the condition on line 52 work?

evanh replied:
The `!=` condition works differently from the `=` condition. We don't add the isNull check to plain `=` conditions.

fpacifici replied:
ok, this is not going to make things worse or more broken, though I would strongly advise for this logic to be revisited be more robust, either now, based on the old query structure, or after the AST is rolled out.

fpacifici reviewed (changes requested):
back to you to fix the condition

evanh committed:
merge

evanh committed:
merge with master

fpacifici reviewed (approved):
did not see further comments around the condition issue.
Let's ensure it gets fixed in a different PR

evanh committed:
Merge branch 'master' into evanh/fix/device_charging_field

evanh committed:
Merge branch 'master' into evanh/fix/device_charging_field

evanh committed:
Merge branch 'master' into evanh/fix/device_charging_field
